### ISSUE TITLE
Maintenance of xpaths (and therefore whole plugin) should be easier

### ISSUE THREAD

**Issue by broeder-j** [JuDFTteam/aiida-fleur#6]
So far all xpaths are hardcodes somewhere in the plugin, therefore the plugin needs a bit of maintenance. 

A solution to extracted all these paths from the schema file, which should not take so much time (or be done only once and installation and stored in a file or so) should replace this.

**Comment by broeder-j** [comment]
maybe most  xpaths should be changed from absolute i.e /fleurinput/.../..../@attribute to //last_node/@attribute. this will eat a bit performance, but might give some more robustness of the xpaths.

**Comment by broeder-j** [comment]
here a reference for xpath syntax: https://www.w3schools.com/xml/xpath_syntax.asp

**Comment by janssenhenning** [comment]
@broeder-j I recently started to look into this and wrote a small function, that can recursively extract the valid xpath to a given tag from a Inputschema. This would essentially be the same as relative paths, but without a performance drop, because the search is done on the rather small InputSchema instead of a big xml file. And I have some more clear ideas, which would be easy to implement

The same can be done with the outputschema, if this is at a point, where one can rely on it to be compatible with the fleur version

**Comment by broeder-j** [comment]
@janssenhenning: years ago I have also written something which does this. through it was to slow. Maybe yours is better.
Also I could not think back then of a way to run this routine only once and not everytime an xpath evaluation is performed (we do lots of those). Back then the schema files where not part of the aiida-fleur repo.

Any ideas here, how to do this once per file now, or only a couple of times?

For parsing the inpxml this should work out fine, it will just replace  ```get_inpxml_file_structure``` and will be run once when any fleurinpData is initialized.

1. But if any function wants for example evaluate the total energy it would have to call a function to give it the xpath for the total energy and if this function first has to open a file to read the paths (i.e we have run the routine on the schema and stored all paths in a file, json or so), a simple string becomes `read it from a file` which I think is to slow.

2. Or we run it once per schema, and hardstore the resulting xpaths in a python file from where we can import them for a certain schema version and provide some utility.  This would only have to be maintained/generated if a new schema file is added, which I think is ok.

3. Another idea is we run this once per schema, and we create a class which has all these xpaths for each schema version, stays in memory and has methods to provide the xpaths. (kind of similar to 2., more complicated, but would need less time to maintain).

Overall, for complex xpath expressions the same problem will remain I guess, because there are the number of xpaths creatable is infinite. But since these are more scarse it becomes maintainable.

I think we could start with 2., because it is quite easy to start and we can propergate it over time through the code and for sure it will be fast enough.

Another thing to consider is that, we would like to make raw parsers for fleur files independent of aiida and move them to masci-tools, which would allow for other non-aiida python utility for fleur to use them (and one could implement a python command line for fleur to parse, convert files etc.). So any implementation has to be modular and needs to allow for disentanglement. Maybe such a feature together with the schema file should be started in masci-tools in the first place.

**Comment by broeder-j** [comment]
@janssenhenning we would of course all be very happy if someone would start and do this ;-)

**Comment by janssenhenning** [comment]
@broeder-j Okay, I will experiment a bit. If I understand 2 correct this file would have some dictionaries with path information and other useful stuff. I think I will first try out the following things and try to generate such a file:

- All possible tag paths (without any complex syntax. If you want something like this you would need to get the last possible tag and construct it from there)
- Attribute paths
   - This also allows for automatic type detection (except float in a couple of places, where xsd:string is used instead of xsd:double)
   - With this one could also automatically generate all supported keys for set_inpchanges
- The same is easily possible for tags_several to ensure that the nothing is overwritten in the inp_dict
-extract sequence information for the places where it is important (e.g. species)

**Comment by janssenhenning** [comment]
The problem with the float attributes that are set to xsd:string is that they are probably needed for manual fleur calculations for using constants
What may be useful for detecting these float attributes is making some alias type for float in the schema which can be double or string so that they are clearly marked

**Comment by janssenhenning** [comment]
@broeder-j Over the weekend I had time to write a small and dirty prototype for the creation of a dictionary with all the information from the fleurinputschema. I pushed it to a new branch in the masci-tools repo
https://github.com/JuDFTteam/masci-tools/blob/fleur_parser/docs/source/index.rst

At the moment I create a file schema_dict.py in the same directory as the inputschema and the desired version is imported via the importlib library from the specified path.

Let me know what you think

**Comment by broeder-j** [comment]
> @broeder-j: Could this also be done for the old inputschemas already in the aiida-fleur git or is that something, that should be avoided?
I think if it is a problem for the old Fleurschema files. Leave it be for now. The only old larger data sets there are, are probably mine and one from Markus, if we want to import that some day.

I probably tried supporting the FLEUR calculation feature, but since there are no tests and nobody has used it it is probably broken. The way to go was to always calculate the values and continue only with the values and no variables.
If this is strongly needed by someone, maybe an a switch to inpgen which will 'evaluate and get rid of all variables' in the inp.xml would also be an option around this.

This 'clearly' marking you want to change in Fleur. The Schema files in AiiDA-fleur should not differ from the fleur ones of course.
Alternatively we can hardcode a list of properties which are 'calculatable' and which would need to be maintained.

**Comment by janssenhenning** [comment]
@broeder-j Okay yes that's something that should be changed in Fleur. I'll open a issue on the fleur git immediately so that we may get this done quickly while they are still finishing the modifications of the inputschema

**Comment by janssenhenning** [comment]
@broeder-j As a first test I made a version of inpxml_todict using the information from the parsed schema dictionary (also on the masci-tools repo). I tested it on the input files in the test directories both of aiida-fleur and the current develop version of fleur and it seems to work very nicely.

Maybe we can discuss at next weeks meeting, what can be improved or added and what to do next

**Comment by broeder-j** [comment]
I think there is no reason to keep the schema files in aiida-fleur.

An idea for a test to see if something should be maintained is: if we run inpgen and we check if the resulting schema written by inpgen matches exactly the one we have for the same version. If we move all schemas to masci tools the tests should be there. Trough I am not sure how to do this in a CI manner.

**Comment by janssenhenning** [comment]
With #111, #115 and #116 merged the bulk of this work is done. Some more places that use explicit xpaths in aiida fleur are listed below:

- [ ] `compress_fleuroutxml()` in `aiida_fleur.tools.io_routines`
- [ ] `extract_corelevels()` in `aiida_fleur.tools.extract_corelevels`
- [ ] several workflows modify the xml file using concrete xpath routines for modifying the input.
   - [ ] SSDisp
   - [ ] DMI
   - [ ] MAE
   - [ ] corehole
- [ ] Several tests outside the test of the now useless `xml_util` use the `eval_xpath` function from that module. This should be replaced by either the masci-tools function or probably easier a pytest fixture

**Comment by broeder-j** [comment]
Nice work!
